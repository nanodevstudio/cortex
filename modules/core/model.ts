export type Model<T> = new () => T;
