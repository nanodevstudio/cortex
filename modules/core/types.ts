import { getModelField } from "./generateSchema";
import { Model } from "./model";

export interface SQLType {
  baseType: string;
  modifiers: string[];
}

export interface FieldType<T, U = T, P extends true | false = false> {
  T: T;
  U: U;
  primary: P;
  references?: {
    model: any;
    column: string;
  };
  sqlType: SQLType;
  columnName?: string;
}

export interface FieldTypeF<T, U = T, P extends true | false = false> {
  (): FieldType<T, U, P>;
}

export const castType = <T>() => (({} as any) as T);

export const makeType = <T>(baseType: string): FieldTypeF<T, T, false> => {
  return () => ({
    T: castType<T>(),
    U: castType<T>(),
    primary: false,
    sqlType: {
      baseType,
      modifiers: ["NOT NULL"],
    },
  });
};

export const text = makeType<string>("text");
export const utcTimestamp = makeType<Date>("timestamp without time zone");
export const integer = makeType<number>("integer");
export const uuid = makeType<string>("uuid");
export const boolean = makeType<boolean>("boolean");
export const serial = makeType<number>("serial");
export const jsonb = makeType<string>("jsonb");

export const array = <T, U, P extends boolean>(
  typeF: FieldTypeF<T, U, P>
): FieldTypeF<T[], U[], P> => {
  return () => {
    const type = typeF();

    return {
      ...type,
      sqlType: {
        ...type.sqlType,
        baseType: `${type.sqlType.baseType}[]`,
      },
    } as any;
  };
};

export const applyModifiers = (type: SQLType, modifiers: string[]): SQLType => {
  return {
    ...type,
    modifiers: [...type.modifiers, ...modifiers],
  };
};

export const primaryGenerated = <T, U, P>(
  typeF: FieldTypeF<T, U, any>
): FieldTypeF<T, U | undefined, true> => {
  return () => {
    const type = typeF();

    return {
      ...type,
      T: type.T,
      primary: true,
      U: castType<typeof type.U | undefined>(),
      sqlType: applyModifiers(type.sqlType, [
        "DEFAULT public.uuid_generate_v4()",
        "PRIMARY KEY",
      ]),
    } as any;
  };
};

export const generatedId = primaryGenerated(uuid);

export const primary = <T, U>(
  fieldType: FieldTypeF<T, U, any>
): FieldTypeF<T, U, true> => {
  return () => ({
    ...fieldType(),
    primary: true,
    sqlType: applyModifiers(fieldType().sqlType, ["PRIMARY KEY"]),
  });
};

export const ref = <T, K extends keyof T>(
  ref: Model<T>,
  columnName: K
): FieldTypeF<
  T[K] extends FieldType<infer T, any, any> ? NonNullable<T> : never,
  T[K] extends FieldType<any, infer U, any> ? NonNullable<U> : never,
  false
> => {
  return () => {
    return {
      T: castType<
        T[K] extends FieldType<infer T, any, any> ? NonNullable<T> : never
      >(),
      U: castType<
        T[K] extends FieldType<any, infer U, any> ? NonNullable<U> : never
      >(),
      primary: false,
      references: {
        model: ref,
        column: columnName as string,
      },
      sqlType: {
        baseType: getModelField(ref, columnName as any)!.sqlType.baseType,
        modifiers: ["NOT NULL"],
      },
    };
  };
};
