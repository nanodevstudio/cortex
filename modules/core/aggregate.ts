import { decodeSelector, DecodeSelector, IDecodeSelector } from "./symbolic";
import { raw, sql, SQLSegment } from "./writes";
import * as uuid from "uuid";
import { getSelector } from "./query";

export class CountAggregation implements IDecodeSelector<number> {
  constructor(public source: DecodeSelector<any[]>) {}

  id = uuid.v4();

  get select() {
    const selector = getSelector(this.source);
    return sql`(select count(*) from ${selector.select} as ${raw(
      JSON.stringify(this.id)
    )})::integer`;
  }
}

export const count = (
  result: DecodeSelector<any[]>
): DecodeSelector<number> => {
  return new CountAggregation(result);
};
