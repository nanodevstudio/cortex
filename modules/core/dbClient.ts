import { Client, ClientConfig } from "pg";

export type DBClient = {
  client?: Client;
  config: ClientConfig;
};

export const makeClient = (config: ClientConfig): DBClient => {
  return {
    config,
  };
};

export const getPGClient = async (client: DBClient) => {
  if (client.client) {
    return client.client;
  }

  const pgClient = new Client(client.config);
  client.client = pgClient;
  await pgClient.connect();


  return pgClient;
};

export const query = async (
  client: DBClient,
  query: string,
  values?: any[]
) => {
  const pg = await getPGClient(client);

  try {
    return await pg.query(query, values);
  } catch (e) {
    throw new Error(
      `SQL Error: ${e.message}, query: \n${query}\n values: ${JSON.stringify(
        values
      )}`
    );
  }
};

export const closeClient = async (client: DBClient) => {
  if (client.client) {
    await client.client.end();
    client.client = undefined;
  }
};
