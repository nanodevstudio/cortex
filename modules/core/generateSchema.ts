import { Model } from "./model";
import { FieldType } from "./types";

const camelToSnakeCase = (str: string) => {
  return (
    str[0].toLowerCase() +
    str.slice(1).replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`)
  );
};

export const getModelInstance = <T>(model: Model<T>) => {
  return new model();
};

export const getModelField = <T>(
  model: Model<T>,
  field: string
): FieldType<any, any, boolean> | undefined => {
  return (getModelInstance(model) as any)[field]();
};

export const getSQLName = <T>(model: Model<T>) => {
  return camelToSnakeCase(model.name);
};

export const getQualifiedSQLTable = <T>(model: Model<T>) => {
  return "public." + JSON.stringify(camelToSnakeCase(model.name));
};

export const getQualifiedSQLColumn = <T>(model: Model<T>, key: string) => {
  return getQualifiedSQLTable(model) + "." + JSON.stringify(key);
};

export const joinLines = (lines: (string | string[])[]) =>
  lines.flat().join("\n");

export const generateForiegnKeys = <T>(model: Model<T>) => {
  const entries = Object.entries(getModelInstance(model));
  return entries
    .filter(([_, field]) => field.references)
    .map(([key, field]) => {
      const { model: refModel, column: refColumn } = field.references;
      const constraintName = JSON.stringify(
        `fk__${getSQLName(model)}_${key}__${getSQLName(refModel)}_${refColumn}`
      );

      return `ALTER TABLE ${getQualifiedSQLTable(
        model
      )} ADD CONSTRAINT ${constraintName} FOREIGN KEY(${JSON.stringify(
        key
      )}) REFERENCES ${getQualifiedSQLTable(refModel)}(${JSON.stringify(
        refColumn
      )})`;
    });
};

export const generateSQLInsert = <T>(model: Model<T>) => {
  const modelName = getQualifiedSQLTable(model);
  const keys = Object.keys(getModelInstance(model));
  const columns = keys.map((key) => {
    return {
      name: key,
      type: getModelField(model, key as any)!.sqlType,
    };
  });

  return joinLines([
    `CREATE TABLE ${modelName} (`,
    columns
      .map(
        ({ name, type }) =>
          `${JSON.stringify(name)} ${type.baseType} ${type.modifiers.join(" ")}`
      )
      .join(",\n"),
    ");",
  ]);
};
