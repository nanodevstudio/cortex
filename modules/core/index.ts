export * as t from "./types";
export * as op from "./operators";
export { select, subselect, DBQuery } from "./query";
export { resetAndSeed, buildSchemaAndSeed } from "./reset";
export type { SeedFn, SeedContext, GlobPath, ResetBasis } from "./reset";
export type { Model } from "./model";
export { makeDBTestManager } from "./postgresManager";
export { insert, remove, update, insertAll, sql } from "./writes";
export type { DBClient } from "./dbClient";
export { makeClient, closeClient } from "./dbClient";
export { count } from "./aggregate";
