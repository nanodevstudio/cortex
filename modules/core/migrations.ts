import { basename } from "path";
import { DBClient, querySQL } from "./dbClient";
import { resolveModulesWithNames } from "./reset";
import { raw, sql } from "./writes";
import * as c from "ansi-colors";

export type MigrationFn = (db: DBClient) => void | Promise<void>;

export interface MigrationOptions {
  db: DBClient;
  migrations: string[];
}

const migrationTableName = "__cortex__migrations";

export const migrate = async ({ db, migrations }: MigrationOptions) => {
  await querySQL(db, sql`BEGIN TRANSACTION;`);

  await querySQL(
    db,
    sql`CREATE TABLE IF NOT EXISTS ${raw(migrationTableName)} (
		id text PRIMARY KEY NOT NULL,
		"transactedAt" timestamp without time zone NOT NULL
	 );`
  );

  const ranMigrations = new Set(
    (
      await querySQL(db, sql`SELECT id FROM ${raw(migrationTableName)};`)
    ).rows.map((value) => value.id)
  );

  const migrationFns = await resolveModulesWithNames<MigrationFn>(migrations);
  const newMigrationFns = migrationFns
    .map((value) => ({ ...value, id: basename(value.path) }))
    .filter((migration) => !ranMigrations.has(migration.id))
    .sort((a, b) => {
      return a.id.localeCompare(b.id);
    });

  for (const migration of newMigrationFns) {
    try {
      await migration.module(db);
      await querySQL(
        db,
        sql`INSERT INTO ${raw(migrationTableName)} VALUES (${migration.id});`
      );
      console.log(c.green(`✅ ran migration: ${migration.id}`));
    } catch (e) {
      console.error(e);
      console.log(c.green(`❌ migration failed, rolledback: ${migration.id}`));
      await querySQL(db, sql`ROLLBACK;`);
      return;
    }
  }

  await querySQL(db, sql`COMMIT;`);
};
