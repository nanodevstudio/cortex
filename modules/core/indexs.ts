import { createHash } from "crypto";
import { getQualifiedSQLTable } from "./generateSchema";
import { Model } from "./model";
import { emptyQuery } from "./query";
import { ModelSymbol, symbolFromQuery } from "./symbolic";
import { getQueryFromSegments, raw, sql, SQLSegment } from "./writes";

export const makeIndex = <M>(
  model: Model<M>,
  opts: (model: ModelSymbol<M>) => SQLSegment
) => {
  const query = emptyQuery(model, { qualify: false });
  const symbol = symbolFromQuery(query);
  const usingQuery = opts(symbol);
  const onQuery = sql`ON ${raw(
    getQualifiedSQLTable(model)
  )} USING ${usingQuery}`;

  const md5 = createHash("md5");
  md5.update(JSON.stringify(getQueryFromSegments(onQuery)));
  const hash = md5.digest("hex");

  const indexQuery = sql`CREATE INDEX IF NOT EXISTS ${raw(
    JSON.stringify(hash)
  )} ${onQuery}`;

  if (model.indexes == null) {
    model.indexes = [];
  }

  model.indexes?.push(indexQuery);

  return {
    id: hash,
  };
};
