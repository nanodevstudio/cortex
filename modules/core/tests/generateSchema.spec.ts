import { generateSQLInsert } from "../generateSchema";
import * as t from "@/core/types";

describe("db/framework/generateSQLInsert()", () => {
  it("can generate sql insert for simple model", () => {
    class Model {
      id = t.generatedId;
      name = t.text;
    }

    expect(generateSQLInsert(Model)).toMatchSnapshot();
  });

  it("can generate sql insert for model with references", () => {
    class Refed {
      id = t.generatedId;
      name = t.text;
    }

    class Model {
      id = t.generatedId;
      name = t.text;
      ref = t.ref(Refed, "id");
    }

    expect(generateSQLInsert(Model)).toMatchSnapshot();
  });
});
