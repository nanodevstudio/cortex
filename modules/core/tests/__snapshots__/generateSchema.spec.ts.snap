// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`db/framework/generateSQLInsert() can generate sql insert for simple model 1`] = `
"CREATE TABLE public.\\"model\\" (
\\"id\\" uuid NOT NULL DEFAULT public.uuid_generate_v4() PRIMARY KEY,
\\"name\\" text NOT NULL
);"
`;
